src/Microsoft.SqlTools.Dmp.Hosting/Protocol/Message.cs

```diff
@@ -0,0 +1,253 @@
+
+        /// <summary>
+        /// Creates a message with an Event type.
+        /// </summary>
+        /// <param name="eventType">Configuration for the event message</param>
+        /// <param name="contents">The contents of the event.</param>
+        /// <typeparam name="TParams"></typeparam>
+        /// <returns>A message with an Event type</returns>
+        public static Message CreateEvent<TParams>(
+            EventType<TParams> eventType, 
+            TParams contents)
+        {
+            JToken contentsToken = contents == null ? null : JToken.FromObject(contents, ContentSerializer);
+            return new Message(MessageType.Event, contentsToken)
+            {
+                Method = eventType.MethodName
+            };
+        }
+        
+        #endregion
+        
+        #region Properties
+        
+        /// <summary>
+        /// Gets or sets the message type.
+        /// </summary>
+        public MessageType MessageType { get; }
+
+        /// <summary>
+        /// Gets or sets the message's sequence ID.
+        /// </summary>
+        public string Id { get; private set; }
+
+        /// <summary>
+        /// Gets or sets the message's method/command name.
+        /// </summary>
+        public string Method { get; private set; }
+
+        /// <summary>
+        /// Gets or sets a JToken containing the contents of the message.
+        /// </summary>
+        public JToken Contents { get; }
+        
+        #endregion
+        
+        #region Serialization/Deserialization
+        
+        public static Message Deserialize(string jsonString)
+        {
+            // Deserialize the object from the JSON into an intermediate object
+            JObject messageObject = JObject.Parse(jsonString);
+            JToken token;
+            
+            // Ensure there's a JSON RPC version or else it's invalid
+            if (!messageObject.TryGetValue("jsonrpc", out token) || token.Value<string>() != JsonRpcVersion)
+            {
+                // TODO: Localize
+                throw new ArgumentException("JSON RPC version parameter is missing or invalid");
+            }
+
+            if (messageObject.TryGetValue("id", out token))
+            {
+                // Message is a Request or Response
+                string messageId = token.ToString();
+
+                if (messageObject.TryGetValue("result", out token))
+                {
+                    return new Message(MessageType.Response, token) {Id = messageId};
+                }
+                if (messageObject.TryGetValue("error", out token))
+                {
+                    return new Message(MessageType.ResponseError, token) {Id = messageId};
+                }
+                
+                JToken messageParams;
+                messageObject.TryGetValue("params", out messageParams);
+                if (!messageObject.TryGetValue("method", out token))
+                {
+                    // TODO: Localize
```

Comment:
Surprised we never localized these? Do we need a separate task to handle doing all that?

Reply:
Actually we never had these checks before. I'll add the localized strings for them.